src/history/hs_rec.c

```diff
@@ -150,16 +150,28 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
               &upd_type_full_diag, existing_val));
             WT_ERR(__wt_compare(session, NULL, existing_val, hs_value, &cmp));
             /*
-             *  Same value should not be inserted again unless 1. previous entry is already
-             * deleted(i.e. the stop timestamp is globally visible), 2. from a different
-             * transaction 3. with a different timestamp if from the same transaction.
+             * The same value should not be inserted again unless:
+             * 1. the previous entry is already deleted (i.e. the stop timestamp is globally
+             * visible)
+             * 2. it came from a different transaction
+             * 3. it came from the same transaction but with a different timestamp
              */
-            if (cmp == 0)
-                WT_ASSERT(session,
-                  __wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw) ||
-                    tw->start_txn == WT_TXN_NONE ||
-                    tw->start_txn != hs_cbt->upd_value->tw.start_txn ||
-                    tw->start_ts != hs_cbt->upd_value->tw.start_ts);
+            if (cmp == 0) {
+                if (!__wt_txn_tw_stop_visible_all(session, &hs_cbt->upd_value->tw) &&
```

Comment:
Can we wrap this in HAVE_DIAGNOSTIC so it is not checked in production build?

Reply:
Yep. This check is already marked in a `HAVE_DIAGNOSTIC`.